src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -12,10 +12,50 @@ namespace Microsoft.CodeAnalysis.Host
 {
     internal partial class TemporaryStorageServiceFactory
     {
+        /// <summary>
+        /// A reference-counting wrapper which allows multiple uses of a single disposable object in code, which is
+        /// deterministically released (by calling <see cref="IDisposable.Dispose"/>) when the last reference is
+        /// disposed.
+        /// </summary>
+        /// <remarks>
+        /// <para>While each instance of <see cref="ReferenceCountedDisposable{T}"/> should be explicitly disposed when
+        /// the object is no longer needed by the code owning the reference, this implementation will not leak resources
+        /// in the event one or more callers fail to do so. The underlying object will be deterministically released
+        /// when the last reference to it is disposed. However, the underlying object is eligible for non-deterministic
+        /// release (should it have a finalizer) when each reference to it is <em>either</em> disposed <em>or</em>
+        /// eligible for garbage collection itself.</para>
+        ///
+        /// <para>All public methods on this type adhere to their pre- and post-conditions and will not invalidate state
+        /// even in concurrent execution. The implementation of <see cref="TryAddReference"/> is lock-free; all other
+        /// methods are wait-free, with the exception of the specific call to <see cref="Dispose"/> which results in the
```

Comment:
why do we need lock-free or wait-free?  It seems to add a *lot* of compelxity to what seems to otherwise just be a case of trying to keep a refcount on a weak reference so that we can release early when the refcount reaches 0.